sstables/mx/reader.cc

```diff
@@ -1217,32 +1227,54 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
     // We avoid unnecessary lookup for single partition reads thanks to this flag
     bool _single_partition_read = false;
     const dht::partition_range& _pr;
-    const query::partition_slice& _slice;
     streamed_mutation::forwarding _fwd;
     mutation_reader::forwarding _fwd_mr;
     read_monitor& _monitor;
+
+    // For reversed (single partition) reads, points to the current position in the sstable
+    // of the reversing data source used underneath (see `partition_reversing_data_source`).
+    // Engaged after `_context` is engaged, i.e. after `initialize()`.
+    const uint64_t* _reversed_read_sstable_position;
 public:
     mx_sstable_mutation_reader(shared_sstable sst,
                             schema_ptr schema,
                             reader_permit permit,
                             const dht::partition_range& pr,
                             const query::partition_slice& slice,
+                            // If `slice` may die after the constructor (but before the read ends),
+                            // `store_slice` tells us to store it.
+                            bool store_slice,
```

Comment:
Why not add a separate constructor which first copies the slice, then delegates to the old constructor.

Reply:
> Why not add a separate constructor which first copies the slice, then delegates to the old constructor.

I'm afraid this is impossible to do in C++ :(
I assume you wanted something like this:
```
    mx_sstable_mutation_reader(shared_sstable sst,
                            schema_ptr schema,
                            reader_permit permit,
                            const dht::partition_range& pr,
                            query::partition_slice&& slice,
                            const io_priority_class& pc,
                            tracing::trace_state_ptr trace_state,
                            streamed_mutation::forwarding fwd,
                            mutation_reader::forwarding fwd_mr,
                            read_monitor& mon)
        : _stored_slice(std::move(slice))
        , mx_sstable_mutation_reader(std::move(sst), std::move(schema), std::move(permit),
                pr, *_stored_slice, pc, std::move(trace_state), fwd, fwd_mr, mon)
    {}
```

Reply:
Yes.

Reply:
In the end I used a single constructor that takes a `value_or_reference<query::partition_slice>`. Check it out.